### PR TITLE
asls: init at 0.4.0

### PR DESCRIPTION
**Pull request by saulecabrera**
###### Motivation for this change
Add a new package for the [AssemblyScript Language Server](https://github.com/saulecabrera/asls)
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/misc/asls/default.nix
Comment (danieldk) [inline pkgs/development/tools/misc/asls/default.nix:5]: Non-standard formatting, use `nixpkgs-fmt` if unsure.

File: pkgs/development/tools/misc/asls/default.nix
Comment (danieldk) [inline pkgs/development/tools/misc/asls/default.nix:14]: Why not build from source?

File: pkgs/development/tools/misc/asls/default.nix
Comment (danieldk) [inline pkgs/development/tools/misc/asls/default.nix:22]:
```suggestion
  install -Dm755 -t $out/bin asls
```

File: pkgs/development/tools/misc/asls/default.nix
Comment (danieldk) [inline pkgs/development/tools/misc/asls/default.nix:28]: Unquoted URLs are deprecated.

```suggestion
    homepage = "https://github.com/saulecabrera/asls";
```

File: pkgs/development/tools/misc/asls/default.nix
Comment (danieldk) [inline pkgs/development/tools/misc/asls/default.nix:23]: I know approximately nothing about Erlang, but the binary also seems to work fine without wrapping? The shebang is patched. Or does embedded code use things from the path?

File: pkgs/development/tools/misc/asls/default.nix
Comment (danieldk) [inline pkgs/development/tools/misc/asls/default.nix:31]: A matter of taste, but I prefer:

```suggestion
    maintainers = with maintainers; [ saulecabrera ];
```

**Review by danieldk — APPROVED** [review]
I have added an initial set of comments. Thanks for your first submission!

**Review by danieldk — CHANGES_REQUESTED** [review]
Didn't mean to approve, sorry ;).

File: pkgs/development/tools/misc/asls/default.nix
Comment (saulecabrera) [reply pkgs/development/tools/misc/asls/default.nix]: You're totally right, having `erlangR22` as part of the `buildInputs` is enought.

File: pkgs/development/tools/misc/asls/default.nix
Comment (saulecabrera) [reply pkgs/development/tools/misc/asls/default.nix:13]: Two main reasons I went with this approach:

- The available tools for the BEAM seem to have several drawbacks not suited for this package:
   - `buildHex`: requires the package to be published in https://hex.pm/ which is not the case for `asls`
   - `buildMix`: requires the package to be a library,` asls `is  an application (BEAM CLI application)
- The other option is to run `mix` commands  during the build phase, that works and that was my initial approach, but that breaks sandboxing (a similar case can be found in this thread: https://discourse.nixos.org/t/question-about-packaging-elixirls/4056)

Until there is a more mature built-in approach I believe this is the most straightforward way.

**Comment by saulecabrera** [comment]
Thanks for your initial comments @danieldk. I've polished the PR, it's ready for another round of reviews.

**Comment by danieldk** [comment]
> Thanks for your initial comments @danieldk. I've polished the PR, it's ready for another round of reviews.

Thanks, I am afraid that I cannot add further comments, since I am not familiar with the Erlang ecosystem. So, I think at this point it's best to wait for someone with Erlang experience to do another review.

**Comment by saulecabrera** [comment]
> So, I think at this point it's best to wait for someone with Erlang experience to do another review.

Agreed.

@DianaOlympos would you have some time for a review? (Apologies for randomly pinging you, but I've noticed that you have been very involved in Nix <> BEAM conversations, if that's not the case, please disregard this message 😄)

**Comment by DianaOlympos** [comment]
I don't see anything problematic here.

Does remind me i need to bring erlang 23 though.

LGTM

File: pkgs/development/tools/misc/asls/default.nix
Comment (timokau) [inline pkgs/development/tools/misc/asls/default.nix:11]: To simplify updates:
```suggestion
    url = "https://github.com/saulecabrera/asls/releases/download/v${version}/bin.tar.gz";
```

**Review by timokau — CHANGES_REQUESTED** [review]
Just one small nit, other than that looks good to me.

File: pkgs/development/tools/misc/asls/default.nix
Comment (saulecabrera) [reply pkgs/development/tools/misc/asls/default.nix]: Good point. Fixed!

Comment (saulecabrera) [reply pkgs/development/tools/misc/asls/default.nix]: I also bumped the version to init at `0.4.0` instead, given that I just released this version today.

**Review by timokau — APPROVED** [review]
Looks good to me now (pending ofBorg's approval), welcome to the team!

**Pull request merged**
